# Unit 12 - Tales from the Crypto

---


## 1. Sentiment Analysis

Use the [newsapi](https://newsapi.org/) to pull the latest news articles for Bitcoin and Ethereum and create a DataFrame of sentiment scores for each coin.

Use descriptive statistics to answer the following questions:
1. Which coin had the highest mean positive score?
2. Which coin had the highest negative score?
3. Which coin had the highest positive score?

In [2]:
# Initial imports
import os
import pandas as pd
from dotenv import load_dotenv, find_dotenv
import nltk as nltk
nltk.download('vader_lexicon')
from nltk.sentiment.vader import SentimentIntensityAnalyzer
analyzer = SentimentIntensityAnalyzer()
from newsapi.newsapi_client import NewsApiClient

%matplotlib inline

[nltk_data] Downloading package vader_lexicon to
[nltk_data]     /Users/tylerhubbard/nltk_data...
[nltk_data]   Package vader_lexicon is already up-to-date!


In [3]:
# Read your api key environment variable
# I need find_dotenv for my computer
load_dotenv(find_dotenv('api_keys.env'))
api_key = os.getenv('NEWS_API_KEY')

In [4]:
# Create a newsapi client
# Make sure to install NewsApiClient (done up above)
newsapi = NewsApiClient(api_key = api_key)

In [5]:
# Fetch the Bitcoin news articles using newsapi and .get_everything
btc_news = newsapi.get_everything(q='bitcoin', language='en')
print(f'Number of relevant Bitcoin news articles in English: {btc_news["totalResults"]}')

Number of relevant Bitcoin news articles in English: 7382


In [6]:
# Fetch the Ethereum news articles using newsapi and .get_everything
eth_news = newsapi.get_everything(q='ethereum', language='en')
print(f'Number of relevant Ethereum news articles in English: {eth_news["totalResults"]}')

Number of relevant Ethereum news articles in English: 3571


In [7]:
# Create the Bitcoin sentiment scores DataFrame using .from_dict()
btc_sentiment = pd.DataFrame.from_dict(btc_news['articles'])
btc_sentiment.head()

source              author  \
0  {'id': 'the-verge', 'name': 'The Verge'}   Elizabeth Lopatto   
1      {'id': None, 'name': 'Slashdot.org'}         EditorDavid   
2      {'id': 'reuters', 'name': 'Reuters'}                None   
3      {'id': 'reuters', 'name': 'Reuters'}                None   
4          {'id': 'wired', 'name': 'Wired'}  Gian M. Volpicelli   

                                               title  \
0  If you’re a Russian YouTuber, how do you get p...   
1                         Why Isn't Bitcoin Booming?   
2  CRYPTOVERSE-Bitcoin could be laid low by miner...   
3  Cryptoverse: Bitcoin gains conflict currency c...   
4  War Is Calling Crypto’s ‘Neutrality’ Into Ques...   

                                         description  \
0  Russian creators are shut off from the global ...   
1  "Bitcoin was seen by many of its libertarian-l...   
2  Bitcoin miners are feeling the heat - and the ...   
3  Bitcoin has leapt since Russia's invasion of U...   
4  War in Ukraine and Western sanctions against R...   

                                                 url  \
0  https://www.theverge.com/2022/3/17/22982122/ru...   
1  https://news.slashdot.org/story/22/03/12/05412...   
2  https://www.reuters.com/markets/europe/cryptov...   
3  https://www.reuters.com/markets/europe/cryptov...   
4  https://www.wired.com/story/crypto-russia-ukra...   

                                          urlToImage           publishedAt  \
0  https://cdn.vox-cdn.com/thumbor/MG_NhB7wSIBIl3...  2022-03-17T13:33:43Z   
1        https://a.fsdn.com/sd/topics/bitcoin_64.png  2022-03-12T18:34:00Z   
2  https://www.reuters.com/resizer/9nBpgfg7pSfpPQ...  2022-02-22T06:17:00Z   
3  https://www.reuters.com/pf/resources/images/re...  2022-03-01T06:10:00Z   
4  https://media.wired.com/photos/6226a83bd53a49d...  2022-03-08T12:00:00Z   

                                             content  
0  When Russia invaded Ukraine, Niki Proshin was ...  
1  "Bitcoin was seen by many of its libertarian-l...  
2  Feb 22 (Reuters) - Bitcoin miners are feeling ...  
3  March 1 (Reuters) - Bitcoin has leapt since Ru...  
4  Whose side is cryptocurrency on? If you had as...

In [8]:
# Create the Ethereum sentiment scores DataFrame
eth_sentiment = pd.DataFrame.from_dict(eth_news['articles'])
eth_sentiment.head()

source  \
0                   {'id': 'wired', 'name': 'Wired'}   
1  {'id': 'business-insider', 'name': 'Business I...   
2           {'id': 'the-verge', 'name': 'The Verge'}   
3                   {'id': 'wired', 'name': 'Wired'}   
4               {'id': None, 'name': 'Entrepreneur'}   

                            author  \
0       Shanti Escalante-De Mattei   
1  prosen@insider.com (Phil Rosen)   
2                Elizabeth Lopatto   
3                  Omar L. Gallaga   
4                     Masha Prusso   

                                               title  \
0       Web3 Threatens to Segregate Our Online Lives   
1  Coinbase earnings show trading of ethereum and...   
2   How Ukrainians are fundraising in cryptocurrency   
3  How People Actually Make Money From Cryptocurr...   
4  What You Need to Know About Ethereum's Role in...   

                                         description  \
0  Governance tokens seem like a tantalizing solu...   
1  Ethereum trading volume increased from 15% to ...   
2  Millions of dollars of cryptocurrency have flo...   
3  Power traders use “staking” and “yield farming...   
4  This now-seven-year-old decentralized and open...   

                                                 url  \
0  https://www.wired.com/story/web3-governance-to...   
1  https://markets.businessinsider.com/news/curre...   
2  https://www.theverge.com/2022/2/26/22952357/uk...   
3  https://www.wired.com/story/how-to-make-money-...   
4        https://www.entrepreneur.com/article/417850   

                                          urlToImage           publishedAt  \
0  https://media.wired.com/photos/621d66c7ea3b8f2...  2022-03-01T14:00:00Z   
1  https://i.insider.com/62190267d0009b001904bd96...  2022-02-25T17:02:30Z   
2  https://cdn.vox-cdn.com/thumbor/teEVxppIZ_JTW-...  2022-02-26T20:29:04Z   
3  https://media.wired.com/photos/622bcc6ef48a924...  2022-03-13T13:00:00Z   
4  https://assets.entrepreneur.com/content/3x2/20...  2022-03-03T16:00:00Z   

                                             content  
0  In February, shit hit the fan in the usual way...  
1  Coinbase reported that the share of trading vo...  
2  Illustration by James Bareham / The Verge\r\n\...  
3  If it sounds too good to be true, youre not wr...  
4  It seems that in 2022, you cant escape from th...

In [9]:
# Create a function with all the results to transform the 'articles' list in a df
# A list of dicts is used to create the df structure, so an empty list 'articles' is created before starting
def create_sentiment_df(articles_list):
    sentiments = []
    
    # Create a for-loop to iterate across the articles list to fetch and filter data
    for article in articles_list['articles']:
        try:
            text = article['content']
            date = article['publishedAt'][:10]
            sentiment = analyzer.polarity_scores(text)
            compound = sentiment['compound']
            neu = sentiment['neu']
            pos = sentiment['pos']
            neg = sentiment['neg']
            
            sentiments.append({
                'text' : text,
                'date' : date,
                'compound' : compound, 
                'neutral' : neu,
                'negative' : neg,
                'positive' : pos
            })
        except AttributeError:
            pass
    df = pd.DataFrame(sentiments)
    return df

In [10]:
# Create btc sentiment df and check it before describing
btc_sentiment_df = create_sentiment_df(btc_news)
btc_sentiment_df.head()

text        date  compound  \
0  When Russia invaded Ukraine, Niki Proshin was ...  2022-03-17    0.0000   
1  "Bitcoin was seen by many of its libertarian-l...  2022-03-12   -0.7713   
2  Feb 22 (Reuters) - Bitcoin miners are feeling ...  2022-02-22   -0.1779   
3  March 1 (Reuters) - Bitcoin has leapt since Ru...  2022-03-01    0.0000   
4  Whose side is cryptocurrency on? If you had as...  2022-03-08   -0.3182   

   neutral  negative  positive  
0    1.000     0.000     0.000  
1    0.831     0.169     0.000  
2    0.887     0.067     0.046  
3    1.000     0.000     0.000  
4    0.854     0.091     0.055

In [11]:
# Describe the Bitcoin Sentiment
btc_sentiment_df.describe()

compound    neutral   negative   positive
count  20.000000  20.000000  20.000000  20.000000
mean    0.001295   0.908500   0.043750   0.047700
std     0.406381   0.071992   0.061198   0.045767
min    -0.771300   0.739000   0.000000   0.000000
25%    -0.231275   0.861500   0.000000   0.000000
50%     0.000000   0.918000   0.000000   0.051500
75%     0.345300   0.952500   0.071000   0.075750
max     0.636900   1.000000   0.187000   0.152000

In [12]:
# Create eth sentiment df and check it before describing
eth_sentiment_df = create_sentiment_df(eth_news)
eth_sentiment_df.head()

text        date  compound  \
0  In February, shit hit the fan in the usual way...  2022-03-01   -0.3182   
1  Coinbase reported that the share of trading vo...  2022-02-25    0.6705   
2  Illustration by James Bareham / The Verge\r\n\...  2022-02-26   -0.4588   
3  If it sounds too good to be true, youre not wr...  2022-03-13    0.8340   
4  It seems that in 2022, you cant escape from th...  2022-03-03   -0.1326   

   neutral  negative  positive  
0    0.848     0.093     0.059  
1    0.812     0.000     0.188  
2    0.917     0.083     0.000  
3    0.713     0.050     0.236  
4    0.956     0.044     0.000

In [13]:
# Describe the Ethereum Sentiment
eth_sentiment_df.describe()

compound   neutral  negative   positive
count  20.000000  20.00000  20.00000  20.000000
mean    0.190925   0.89145   0.02975   0.078750
std     0.401422   0.10092   0.04006   0.085383
min    -0.526700   0.69200   0.00000   0.000000
25%     0.000000   0.84100   0.00000   0.000000
50%     0.025800   0.91050   0.00000   0.066500
75%     0.514625   1.00000   0.06125   0.120500
max     0.834000   1.00000   0.11500   0.249000

### Questions:

Q: Which coin had the highest mean positive score?

A: Ethereum (0.0788) had a higher score than Bitcoin (0.0477).

Q: Which coin had the highest compound score?

A: Ethereum's max score (0.834) had a higher score than Bitcoin (0.637)

Q. Which coin had the highest positive score?

A: Ethereum had a higher positive score (0.249) than Bitcoin (0.152)

---

## 2. Natural Language Processing
---
###   Tokenizer

In this section, you will use NLTK and Python to tokenize the text for each coin. Be sure to:
1. Lowercase each word.
2. Remove Punctuation.
3. Remove Stopwords.

In [30]:
from nltk.tokenize import word_tokenize, sent_tokenize
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer, PorterStemmer
from string import punctuation
import re

In [53]:
# I have to use the following for my stopwords:
import nltk
nltk.download('stopwords')
nltk.download('punkt')
nltk.download('wordnet')
nltk.download('omw-1.4')

[nltk_data] Downloading package stopwords to
[nltk_data]     /Users/tylerhubbard/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package punkt to
[nltk_data]     /Users/tylerhubbard/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package wordnet to
[nltk_data]     /Users/tylerhubbard/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package omw-1.4 to
[nltk_data]     /Users/tylerhubbard/nltk_data...
[nltk_data]   Unzipping corpora/omw-1.4.zip.


True

In [32]:
# Instantiate the lemmatizer
lemmatizer = WordNetLemmatizer()

# Create a list of stopwords
stop_words = set(stopwords.words('english'))

# Expand the default stopwords list if necessary
stop_words = stop_words.union({'reuters', 'com', 'char', 'ha', 'illustration'})

In [55]:
# Complete the tokenizer function
def tokenizer(text):
    """Tokenizes text."""
    
    # Remove the punctuation from text
    text = text.translate(str.maketrans('', '', punctuation))
   
    # Create a tokenized list of the words
    tokens_re = word_tokenize(text)
    
    # Lemmatize words into root words
    lemmatized_tokens = [lemmatizer.lemmatize(token) for token in tokens_re]

   
    # Convert the words to lowercase and remove stops
    tokens = [token.lower() for token in lemmatized_tokens if token.lower() not in stop_words]
              
    return tokens

In [60]:
# Add a new tokens column to a df
def addTokensColumn(df_tokens_column):
    
    # Make a .copy() to reuse
    df = df_tokens_column.copy()
    
    # Create the column
    df['tokens'] = df.apply(lambda row: (tokenizer(row.text)), axis=1)
    cols = ['date', 'text', 'tokens', 'compound', 'positive', 'negative', 'neutral']
    df = df[cols]
    
    return df

In [61]:
# Create a new tokens column for Bitcoin
btc_sentiment_df = addTokensColumn(btc_sentiment_df)
btc_sentiment_df.head()

date                                               text  \
0  2022-03-17  When Russia invaded Ukraine, Niki Proshin was ...   
1  2022-03-12  "Bitcoin was seen by many of its libertarian-l...   
2  2022-02-22  Feb 22 (Reuters) - Bitcoin miners are feeling ...   
3  2022-03-01  March 1 (Reuters) - Bitcoin has leapt since Ru...   
4  2022-03-08  Whose side is cryptocurrency on? If you had as...   

                                              tokens  compound  positive  \
0  [russia, invaded, ukraine, niki, proshin, wa, ...    0.0000     0.000   
1  [bitcoin, wa, seen, many, libertarianleaning, ...   -0.7713     0.000   
2  [feb, 22, bitcoin, miner, feeling, heat, pain,...   -0.1779     0.046   
3  [march, 1, bitcoin, leapt, since, russias, inv...    0.0000     0.000   
4  [whose, side, cryptocurrency, asked, satoshi, ...   -0.3182     0.055   

   negative  neutral  
0     0.000    1.000  
1     0.169    0.831  
2     0.067    0.887  
3     0.000    1.000  
4     0.091    0.854

In [62]:
# Create a new tokens column for Ethereum
eth_sentiment_df = addTokensColumn(eth_sentiment_df)
eth_sentiment_df.head()

date                                               text  \
0  2022-03-01  In February, shit hit the fan in the usual way...   
1  2022-02-25  Coinbase reported that the share of trading vo...   
2  2022-02-26  Illustration by James Bareham / The Verge\r\n\...   
3  2022-03-13  If it sounds too good to be true, youre not wr...   
4  2022-03-03  It seems that in 2022, you cant escape from th...   

                                              tokens  compound  positive  \
0  [february, shit, hit, fan, usual, way, old, tw...   -0.3182     0.059   
1  [coinbase, reported, share, trading, volume, e...    0.6705     0.188   
2  [james, bareham, verge, 15, million, donated, ...   -0.4588     0.000   
3  [sound, good, true, youre, wrong, yield, farmi...    0.8340     0.236   
4  [seems, 2022, cant, escape, metaverse, faceboo...   -0.1326     0.000   

   negative  neutral  
0     0.093    0.848  
1     0.000    0.812  
2     0.083    0.917  
3     0.050    0.713  
4     0.044    0.956

---

### NGrams and Frequency Analysis

In this section you will look at the ngrams and word frequency for each coin. 

1. Use NLTK to produce the n-grams for N = 2. 
2. List the top 10 words for each coin. 

In [15]:
from collections import Counter
from nltk import ngrams

In [16]:
# Generate the Bitcoin N-grams where N=2
# YOUR CODE HERE!

In [17]:
# Generate the Ethereum N-grams where N=2
# YOUR CODE HERE!

In [18]:
# Function token_count generates the top 10 words for a given coin
def token_count(tokens, N=3):
    """Returns the top N tokens from the frequency count"""
    return Counter(tokens).most_common(N)

In [19]:
# Use token_count to get the top 10 words for Bitcoin
# YOUR CODE HERE!

In [20]:
# Use token_count to get the top 10 words for Ethereum
# YOUR CODE HERE!

---

### Word Clouds

In this section, you will generate word clouds for each coin to summarize the news for each coin

In [21]:
from wordcloud import WordCloud
import matplotlib.pyplot as plt
plt.style.use('seaborn-whitegrid')
import matplotlib as mpl
mpl.rcParams['figure.figsize'] = [20.0, 10.0]

In [22]:
# Generate the Bitcoin word cloud
# YOUR CODE HERE!

In [23]:
# Generate the Ethereum word cloud
# YOUR CODE HERE!

---
## 3. Named Entity Recognition

In this section, you will build a named entity recognition model for both Bitcoin and Ethereum, then visualize the tags using SpaCy.

In [24]:
import spacy
from spacy import displacy

In [25]:
# Download the language model for SpaCy
# !python -m spacy download en_core_web_sm

In [26]:
# Load the spaCy model
nlp = spacy.load('en_core_web_sm')

---
### Bitcoin NER

In [27]:
# Concatenate all of the Bitcoin text together
# YOUR CODE HERE!

In [28]:
# Run the NER processor on all of the text
# YOUR CODE HERE!

# Add a title to the document
# YOUR CODE HERE!

In [29]:
# Render the visualization
# YOUR CODE HERE!

In [30]:
# List all Entities
# YOUR CODE HERE!

---

### Ethereum NER

In [31]:
# Concatenate all of the Ethereum text together
# YOUR CODE HERE!

In [32]:
# Run the NER processor on all of the text
# YOUR CODE HERE!

# Add a title to the document
# YOUR CODE HERE!

In [33]:
# Render the visualization
# YOUR CODE HERE!

In [34]:
# List all Entities
# YOUR CODE HERE!

---